labels: Keep LabelArrays sorted

jrajahalme opened the pull request:
Create `labels.LabelArray` in sorted order which allows interacting with KV-store without worrying about label order. To be safe, all other uses of LabelArrays are also changed to sorted order, which affects e.g., rule labels. Arguably this is better than the somewhat arbitrary order they were in before.

This PR makes the follow-up work on `policy.SelectorCache` (#7539) easier. This also allows use of faster matching algorithms on LabelArrays in future.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7815)
<!-- Reviewable:end -->

aanm commented on pkg/k8s/json_patch.go:
```diff
@@ -14,7 +14,7 @@
 package k8s
 
-const(
+const (
```
There's a PR already for this change https://github.com/cilium/cilium/pull/7814

jrajahalme replied:
dropping this once that merges

jrajahalme replied:
dropped this for the current version, as CI does not care about this :-)

aanm commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -53,11 +53,12 @@ var (
 
 // GetPolicyLabels returns a LabelArray for the given namespace and name.
 func GetPolicyLabels(ns, name string, uid types.UID, derivedFrom string) labels.LabelArray {
```
please add a unit test for this function to make sure the order forever is correct,

jrajahalme replied:
There are many already, adjusted them accordingly.

aanm commented on pkg/k8s/network_policy_test.go:
```diff
@@ -1640,10 +1640,10 @@ func (s *K8sSuite) TestGetPolicyLabelsv1(c *C) {
 		c.Assert(lbls, NotNil)
 		c.Assert(len(lbls), Equals, 4, Commentf(
 			"Incorrect number of labels: Expected Name, UID, Namespace and DerivedFrom labels."))
```
You need to fix the comment

jrajahalme replied:
Changed the order in the error message, too.

aanm commented on pkg/k8s/network_policy_test.go:
```diff
@@ -1640,10 +1640,10 @@ func (s *K8sSuite) TestGetPolicyLabelsv1(c *C) {
 			"Incorrect number of labels: Expected Name, UID, Namespace and DerivedFrom labels."))
-		assertLabel(lbls[0], k8sConst.PolicyLabelName, tt.name)
-		assertLabel(lbls[1], k8sConst.PolicyLabelUID, tt.uuid)
+		assertLabel(lbls[0], k8sConst.PolicyLabelDerivedFrom, tt.derivedFrom)
```
Do not use constants here, use their values instead, if we ever change the value of the constant we might change the order accidentally.

jrajahalme replied:
Changed.

jrajahalme committed:
identity: Support creating a new Identity with a LabelArray

Add NewIdentityFromLabelArray() to create a new Identity when a sorted
LabelArray already exists, e.g., when identity is received from a
KV-store.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
labels: Support creating LabelArrays directly.

Add support for creating LabelArrays from SortedList and turning a
LabelArray to Labels. This helps avoiding unnecessary work currently
done when creating Labels from SortedList and then LabelArray from the
Labels.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
labels: Always produce a sorted LabelArray()

LabelArray users care about the order, eg., when producing a KV-store
key. Never produce unsorted LabelArrays, so that users can depend on
the array to be already sorted.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
identity: Change globalIdentity to wrap a LabelArray

Change the identity/cache package internal type 'globalIdentity' to
wrap a (sorted) LabelArray rather than (unsorted) Labels, as that is a
better fit for (always sorted) KV-store keys.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
